xdcr.go

```diff
@@ -0,0 +1,245 @@
+
+	}
+
+	for _, dsName := range dataStores {
+		dataStore, err := r.fromBucket.NamedDataStore(dsName)
+		if err != nil {
+			return nil, fmt.Errorf("Could not get data store %s: %w", dsName, err)
+		}
+
+		if dataStore.GetCollectionID() == collectionID {
+			name := dataStore.DataStoreName()
+			r.fromBucketCollectionIDs[collectionID] = name
+			return name, nil
+
+		}
+	}
+	return nil, fmt.Errorf("Could not find collection with ID %d", collectionID)
+}
+
+// Start starts the replication.
+func (r *XDCR) Start(ctx context.Context) error {
+	args := sgbucket.FeedArguments{
+		ID:         "xdcr-" + r.replicationID,
+		Backfill:   sgbucket.FeedNoBackfill,
+		Terminator: r.terminator,
+	}
+
+	callback := func(event sgbucket.FeedEvent) bool {
+		docID := string(event.Key)
+		trace("Got event %s, opcode: %s", docID, event.Opcode)
+		dsName, err := r.getFromBucketCollectionName(event.CollectionID)
+		if err != nil {
+			warn("Could not find collection with ID %d for docID %s: %s", event.CollectionID, docID, err)
+			r.errorCount.Add(1)
+			return false
+
+		}
+
+		switch event.Opcode {
+		case sgbucket.FeedOpDeletion, sgbucket.FeedOpMutation:
+			if strings.HasPrefix(docID, sgbucket.SyncDocPrefix) && !strings.HasPrefix(docID, sgbucket.Att2Prefix) {
+				trace("Filtering doc %s", docID)
+				r.docsFiltered.Add(1)
+				return true
+
+			}
+
+			toDataStore, err := r.toBucket.NamedDataStore(dsName)
+			if err != nil {
+				warn("Replicating doc %s, could not find matching datastore for %s in target bucket", event.Key, dsName)
+				r.errorCount.Add(1)
+				return false
+			}
+
+			originalCas, err := toDataStore.Get(docID, nil)
+			if err != nil && !toDataStore.IsError(err, sgbucket.KeyNotFoundError) {
+				warn("Skipping replicating doc %s, could not perform a kv op get doc in toBucket: %s", event.Key, err)
+				r.errorCount.Add(1)
+				return false
+			}
+
+			/* full LWW conflict resolution is not implemented in rosmar yet
+
+			CBS algorithm is:
+
+			if (command.CAS > document.CAS)
+			  command succeeds
+			else if (command.CAS == document.CAS)
+			  // Check the RevSeqno
+			  if (command.RevSeqno > document.RevSeqno)
+			    command succeeds
+			  else if (command.RevSeqno == document.RevSeqno)
+			    // Check the expiry time
+			    if (command.Expiry > document.Expiry)
+			      command succeeds
+			    else if (command.Expiry == document.Expiry)
+			      // Finally check flags
+			      if (command.Flags < document.Flags)
+			        command succeeds
+
+
+			command fails
+
+			In the current state of rosmar:
+
+			1. all CAS values are unique.
+			2. RevSeqno is not implemented
+			3. Expiry is implemented and could be compared except all CAS values are unique.
+			4. Flags are not implemented
+
+			*/
+
+			if event.Cas <= originalCas {
+				trace("Skipping replicating doc %s, cas %d <= %d", docID, event.Cas, originalCas)
+				return true
+			}
+
+			toCollection, ok := toDataStore.(*Collection)
```

Comment:
It's unexpected to me that we're doing this check after we've already done toDataStore.Get.  Can we just get the collection in the first place on line 107?

Reply:
I changed this code to avoid casts by implementing fixes in sgbucket to consolidate these interfaces into DataStore. This code exists in a separate PR but is included in this PR.